src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1159,12 +1163,24 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
                     nameof(expression));
             }
 
+            var inputTypeString = GetInputTypeString(inputType);
             var tagBuilder = new TagBuilder("input");
             tagBuilder.TagRenderMode = TagRenderMode.SelfClosing;
             tagBuilder.MergeAttributes(htmlAttributes);
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            var suppliedTypeString = tagBuilder.Attributes["type"];
+            if (_placeholderInputTypes.Contains(suppliedTypeString))
+            {
+                modelExplorer = modelExplorer ?? ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+                var placeholder = modelExplorer.Metadata.Placeholder;
+                if (placeholder != null)
```

Comment:
@tuespetre shouldn't this be `string.IsNullOrEmpty(placeholder)`? I.e. when is it useful to generate `placeholder=""`?